README.md
<p align="center"><img src="https://socialify.git.ci/stfnhmplr/homebridge-synology/image?description=1&font=Inter&issues=1&language=1&pattern=Charlie%20Brown&stargazers=1&theme=Light" alt="project"></p>

## 🧐 Features
- Wake up (WOL has to be enabled) and shutdown your Synology Diskstation
- Get the current system temperature
- Supports 2-Factor-Authentication
- Configuration through homebridge-ui-x

**If you would like to support me or the further development, please consider buying me a coffee.**

<a href="https://www.buymeacoffee.com/himpler" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" style="height: 60px !important;width: 217px !important;" ></a>

## 🚀 Installation
<p>Follow the instruction in NPM for the homebridge server installation. The plugin is published through NPM and should be installed "globally" by typing:</p>

```
sudo npm install -g homebridge-synology
```

## 🛠️ Configuration
Edit your `config.json` and add a new accessory. Example:

```json
{
    "bridge": {
        "name": "Homebridge",
        "username": "CC:22:3D:E3:CE:51",
        "port": 51826,
        "pin": "031-45-154"
    },
    "description": "This is an example configuration file for the homebridge synology plugin",
    "hint": "Always paste into jsonlint.com validation page before starting your homebridge, saves a lot of frustration",
    "accessories": [
        {
            "accessory": "synology",
            "name": "Diskstation",
            "host": "192.168.1.1",
            "mac": "A1:B2:C3:D4:E5:F6",
            "port": 5000,
            "protocol": "http",
            "username": "your-username",
            "password": "your-password",
            "version": "6.2.2",
            "otp": "otp-code for 2FA",
            "startupTime": 60,
            "shutdownTime": 60,
            "disabled": [],
        }
    ]
}
```
### Some explanations
- **Version:** Your current DSM Version. **Important:** If you are using DSM version > 6.2.2 or DSM 7, enter `6.2.2` here anyway.
- **OTP (optional):** If you have enabled 2-Factor-Authentication, the code must be entered here. For more information, see https://github.com/iobroker-community-adapters/ioBroker.synology/blob/HEAD/docs/en/template.md
- **Startup and shutdown time (optional):** You can specify a duration for the startup and the shutdown process. During this time, there is no status change due to polling. Both defaults to 60s.
- **disabled (optional):** You can disable features. The services to be deactivated must be specified as an array of strings, such as `["switch", "temperature", "diskTemperature"]`. If you disable the switch functionality, you can't start or stop your diskstation anymore.

## 🛡️ License
This project is licensed under the MIT


CHANGELOG.md
# Changelog

## [0.5.0] - 2021-12-16
- added average disk temperature

## [0.4.1] - 2021-10-13
- fixed typo in config schema
- fixed protocol configuration
- updated readme

## [0.4.0] - 2021-10-12
- totally rewritten in typescript
- added support for 2FA


index.js
var Service, Characteristic;
var Synology = require('./lib/synology');
var inherits = require('util').inherits;
var pollingtoevent = require('polling-to-event');


module.exports = function (homebridge) {
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;

    fixInheritance(SynologyAccessory.CpuLoad, Characteristic);
    fixInheritance(SynologyAccessory.DiskUsage, Characteristic);
    fixInheritance(SynologyAccessory.StatsService, Service);

    homebridge.registerAccessory('homebridge-synology', 'Synology', SynologyAccessory);
};


function fixInheritance(subclass, superclass) {
    var proto = subclass.prototype;
    inherits(subclass, superclass);
    subclass.prototype.parent = superclass.prototype;
    for (var mn in proto) {
        subclass.prototype[mn] = proto[mn];
    }
}


function SynologyAccessory(log, config) {
    this.log = log;
    this.config = config;
    this.name = config.name;

    this.log('Diskstation url: http' + (config.secure ? 's' : '') + '://' + config.ip + ':' + config.port);

    this.synology = new Synology({
      ip: config.ip,
      mac: config.mac,
      secure: config.secure || null,
      port: config.port || null,
      version: config.version,
      user: config.user || config.account,
      passwd: config.password,
      timeout: config.timeout || 3000
    });

    var that = this;

    this.doPolling = config.doPolling || false;
	this.pollingInterval = config.pollingInterval || 60;
	this.pollingInterval = parseInt(this.pollingInterval);

	this.setAttempt = 0;
	this.state = false;

	if (this.pollingInterval < 10 && this.pollingInterval > 100000) {
		this.log('polling interval out of range... disabled polling');
		this.doPolling = false;
	}

	// Status Polling
	if (this.doPolling) {
		that.log('start polling...');
		var statusemitter = pollingtoevent(function(done) {
			that.log('do poll...')
			that.getPowerState(function(error, state) {
				done(error, state, that.setAttempt);
			}, 'statuspoll');
		},{
		    longpolling: true,
		    interval: that.pollingInterval * 1000,
		    longpollEventName:'statuspoll'
		});

		statusemitter.on('statuspoll', function(data) {
			that.state = data;
			that.log('poll end, state: ' + data);

			if (that.switchService ) {
			    that.switchService.getCharacteristic(Characteristic.On)
    			.updateValue(that.state, null, 'statuspoll');
			}
		});
	}

}


SynologyAccessory.CpuLoad = function () {
    Characteristic.call(this, 'CPU Load', '12d21a89-9466-4548-8edd-b05e6b93c23e');
    this.setProps({
        format: Characteristic.Formats.UINT8,
        unit: Characteristic.Units.PERCENTAGE,
        perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
    });
    this.value = this.getDefaultValue();
};


SynologyAccessory.DiskUsage = function () {
    Characteristic.call(this, 'Disk Usage', 'de3c3d3d-6f86-446c-9dac-535858736ddd');
    this.setProps({
        format: Characteristic.Formats.UINT8,
        unit: Characteristic.Units.PERCENTAGE,
        perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
    });
    this.value = this.getDefaultValue();
};


SynologyAccessory.StatsService = function (displayName, subtype) {
    Service.call(this, displayName, '9d0ea4eb-31db-47e9-83ef-302193e669d8', subtype);
    this.addCharacteristic(new SynologyAccessory.DiskUsage());
    this.addCharacteristic(new SynologyAccessory.CpuLoad());
};


SynologyAccessory.prototype.getPowerState = function (callback, context) {
    var that = this;

    if ((!context || context != 'statuspoll') && this.doPolling) {
		callback(null, this.state);
	} else {
        that.synology.getPowerState(function (err, state) {
            if (!err) {
                that.log('current power state is: ' + state);
                callback(null, state);
            } else {
                that.log(err);
                callback(err);
            }
        });
    }
};


SynologyAccessory.prototype.setPowerState = function (powerState, callback, context) {
    var that = this;

    //don't set the value while polling
	if (context && context === 'statuspoll') {
		callback(null, powerState);
	    return;
	}

    this.setAttempt++;

    if (powerState) { //turn on
        that.synology.wakeUp(function (err) {
            if (!err) {
                that.log('Diskstation woke up!');
                callback(null);
            } else {
                that.log('Something went wrong: ' + err);
                callback(err);
            }
        });
    }

    else { //turn off
        that.synology.shutdown(function (err) {
            if (!err) {
                that.log('Shutting down Diskstation')
                callback(null);
            } else {
                that.log('Error shutting down Diskstation: ' + err)
                callback(err);
            }
        });
    }
};


SynologyAccessory.prototype.getCpuLoad = function (callback) {
    var that = this;

    if(!that.state) {
        callback(null, 0)
        return;
    }

    that.synology.getCpuLoad(function (err, data) {
        if (!err) {
            that.log('current cpu load: %s %', data);
            callback(null, data);
        } else {
            that.log(err);
            callback(null, 0); //testing
        }
    });
};


SynologyAccessory.prototype.getDiskUsage = function (callback) {
    var that = this;

    if(!that.state) {
        callback(null, 0)
        return;
    }

    that.synology.getDiskUsage(function (err, data) {
        if (!err) {
            that.log('current volume usage: %s %', data);
            callback(null, data);
        } else {
            that.log(err);
            callback(null, 0); //testing
        }
    });
};


SynologyAccessory.prototype.getSystemTemp = function (callback) {
    var that = this;

    if(!that.state) {
        callback(null, 0)
        return;
    }

    that.synology.getSystemTemp(function (err, data) {
        if (!err) {
            that.log('current system temp: %s °C', data);
            callback(null, data);
        } else {
            that.log(err);
            callback(null, 0); //testing
        }
    });
};


SynologyAccessory.prototype.getServices = function () {
    var informationService = new Service.AccessoryInformation();

    informationService
        .setCharacteristic(Characteristic.Name, this.name)
        .setCharacteristic(Characteristic.Manufacturer, 'Synology');

    this.switchService = new Service.Switch(this.name);
    this.switchService.getCharacteristic(Characteristic.On)
        .on('get', this.getPowerState.bind(this))
        .on('set', this.setPowerState.bind(this));

    var statsService = new SynologyAccessory.StatsService(this.name + ' status');
    statsService.getCharacteristic(SynologyAccessory.DiskUsage)
        .on('get', this.getDiskUsage.bind(this));
    statsService.getCharacteristic(SynologyAccessory.CpuLoad)
        .on('get', this.getCpuLoad.bind(this));

    var tempService = new Service.TemperatureSensor(this.name + ' temperature');
    tempService.getCharacteristic(Characteristic.CurrentTemperature)
        .on('get', this.getSystemTemp.bind(this));

    var services = [informationService];

    if ('disabled' in this.config) {
        if (this.config.disabled.indexOf("switch") === -1) { services.push(this.switchService); }
        if (this.config.disabled.indexOf("stats") === -1) { services.push(statsService); }
        if (this.config.disabled.indexOf("temp") === -1) { services.push(tempService); }
    } else {
        var services = [informationService, this.switchService, tempService, statsService];
    }

    return services;
};


lib/synology.js
/**
 * synology.js functions for homebridge
 * 12.03.2016 by stfnhmplr
 */

var request = require('request');
var wol = require('wake_on_lan');

var Synology = function(params) {
    this.ip = params.ip;
    this.mac = params.mac;
    this.secure = params.secure || false;
    this.port = params.port || (this.secure ? 5001 : 5000);
    this.timeout = parseInt(params.timeout) || 5000; //request timeout
    this.version = parseInt(params.version) ||  6;

    this.user = params.user;
    this.passwd = params.passwd;
    this.auth = {
        sid: '', //session id
        time: '', //unix time
        timeout: 15 * 60 //in sec
    };

    this.url = 'http' + (this.secure ? 's' : '') + '://' + this.ip + ':' + this.port;
};


/**
 * check if the sid is still valid
 * @returns {string|boolean}
 */
Synology.prototype.isLoggedIn = function() {
    return (this.auth.sid && (this.auth.time + this.auth.timeout) > (new Date / 1e3 | 0)) ? true : false;
};


/**
 * Login to your diskstation
 * @return {Promise}
 */
Synology.prototype._login = function() {
    var that = this;

    return new Promise(function(resolve, reject) {
        if(that.isLoggedIn()) {
            resolve("Still logged in")
        } else {
            var options = {
                url: that.url + '/webapi/auth.cgi',
                timeout: that.timeout,
                qs: {
                    api: 'SYNO.API.Auth',
                    method: 'login',
                    version: 3,
                    account: that.user,
                    passwd: that.passwd,
                    session: 'homebridge-synology-' + Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 10),
                    format: 'sid'
                },
                rejectUnauthorized: false,
                strictSSL: false
            };

            request.get(options, function(err, res, body) {
                if(!err && res.statusCode === 200) {
                    var json = JSON.parse(body);
                    if ('error' in json) {
                        reject("Can't login to Diskstation. Error Code: " + json.error.code)
                    } else {
                        that.auth.sid = json.data.sid;
                        that.auth.time = (new Date / 1e3 | 0);
                        resolve('Login successfull');
                    }
                } else {
                    reject("Can't login to Diskstation. " + err)
                }
            });
        }
    });
};


/**
 * get the power State of your Diskstation
 * @param callback
 */
Synology.prototype.getPowerState = function(callback) {
    that = this;

    var options = {
        url: that.url + '/webman/index.cgi',
        method: 'GET',
        timeout: that.timeout,
        rejectUnauthorized: false,
        strictSSL: false
    }

    request(options, function(err, res) {
        if (!err && res.statusCode === 200) {
            callback(null, true);
        } else {
            callback(null, false);
        }
    });
};


/**
 * Wake on LAN support for Diskstation
 * wol has to be enabled
 * @param callback
 */
Synology.prototype.wakeUp = function(callback) {
    var that = this;

    wol.wake(that.mac, function(err) {
        if (!err) {
            callback(null)
        } else {
            callback(err);
        }
    });
};


/**
 * Shutdown your Diskstation
 * @param callback
 */
Synology.prototype.shutdown = function(callback) {
    var that = this;

    that._login().then(function() {
        var apiUrl = (that.version >= 6) ? '/webapi/entry.cgi' : '/webapi/dsm/system.cgi';

        var options = {
            url: that.url + apiUrl,
            qs: {
                api: (that.version >= 6) ? 'SYNO.Core.System' : 'SYNO.DSM.System',
                version: 1,
                method: 'shutdown',
                _sid: that.auth.sid
            },
            method: 'GET',
            rejectUnauthorized: false,
            strictSSL: false
        };

        request(options, function(err, res, body) {
            (!err && res.statusCode === 200 && JSON.parse(body).success) ? callback(null): callback(err);
        });
    }).catch(function(err) {
        callback(err);
    });

};


/**
 * gets the average disk temperature
 * @param  {Function} callback [description]
 * @return {[type]}            [description]
 */
Synology.prototype.getDiskTemp = function(callback) {
    var that = this;

    that._login().then(function(res) {
        var apiUrl = (that.version >= 6) ? '/webapi/entry.cgi' : '/webapi/dsm/info.cgi';
        var options = {
            url: that.url + apiUrl,
            qs: {
                api: (that.version >= 6) ? 'SYNO.Core.System' : '',
                version: 1,
                method: (that.version >= 6) ? 'info' : 'getinfo',
                type: 'storage',
                _sid: that.auth.sid
            },
            method: 'GET',
            rejectUnauthorized: false,
            strictSSL: false
        };

        request(options, function(err, res, body) {
            if (!err && res.statusCode === 200) {
                var json = JSON.parse(body);
                if (json.success) {
                    var temp = 0;
                    for (var i = json.data.hdd_info.length; i--;) {
                        temp += json.data.hdd_info[i].temp;
                    }
                    temp = Math.round(temp / json.data.hdd_info.length);
                    callback(null, temp);
                } else {
                    callback(err);
                }
            } else {
                callback(err);
            }
        });
    }).catch(function(err) {
        callback(err);
    });
}


/**
 * System temperature of your diskstation
 * If not available, it returns your average disk temperature
 * @param callback
 */
Synology.prototype.getSystemTemp = function(callback) {
    var that = this;

    that._login().then(function() {
        var apiUrl = (that.version >= 6) ? '/webapi/entry.cgi' : '/webapi/dsm/info.cgi';
        var options = {
            url: that.url + apiUrl,
            qs: {
                api: (that.version >= 6) ? 'SYNO.Core.System' : 'SYNO.DSM.Info',
                version: 1,
                method: (that.version >= 6) ? 'info' : 'getinfo',
                _sid: that.auth.sid
            },
            method: 'GET',
            rejectUnauthorized: false,
            strictSSL: false
        };
        request(options, function(err, res, body) {
            if (!err && res.statusCode == 200) {
                var json = JSON.parse(body);
                if (json.success && typeof json.data.temperature !== 'undefined') {
                    callback(null, json.data.temperature);
                } else {
                    that.getDiskTemp(function(error, data) {
                        if(!error) {
                            callback(null, data)
                        } else {
                            callback("An error occured while getting SystemTemp: " + error)
                        }
                    });
                }
            } else {
                callback("An error occured while getting SystemTemp: " + err);
            }
        });
    }).catch(function(err) {
        callback("An error occured while getting SystemTemp: " + err);
    });
};


/**
 * returns the current cpu load
 * @param callback
 */
Synology.prototype.getCpuLoad = function(callback) {
    var that = this;

    that._login().then(function() {

        var apiUrl = (that.version >= 6) ? '/webapi/entry.cgi' : '/webapi/dsm/system_loading.cgi';
        var options = {
            url: that.url + apiUrl,
            qs: {
                api: (that.version >= 6) ? 'SYNO.Core.System.Utilization' : 'SYNO.DSM.SystemLoading',
                version: 1,
                method: (that.version >= 6) ? 'get' : 'getinfo',
                type: 'current',
                resource: ['cpu'],
                _sid: that.auth.sid
            },
            method: 'GET',
            rejectUnauthorized: false,
            strictSSL: false
        };

        request(options, function(err, res, body) {
            if (!err && res.statusCode == 200) {
                var json = JSON.parse(body);
                if (json.success) {
                    if (that.version >= 6) {
                        var load = json.data.cpu['other_load'] +
                            json.data.cpu['system_load'] +
                            json.data.cpu['user_load'];
                    } else {
                        load = Math.round(json.data.cpu.user * 100);
                    }
                    callback(null, load);
                } else {
                    callback("An error occured while getting CpuLoad: " + json.error.code);
                }
            } else {
                callback("An error occured while getting CpuLoad: " + err);
            }
        });
    }).catch(function(err) {
        callback("An error occured while getting CpuLoad: " + err);
    })
};


/**
 * Gets the current disk/volume usage quote
 * returns the average if there are more than one volume
 * @param callback
 */
Synology.prototype.getDiskUsage = function(callback) {
    var that = this;

    that._login().then(function() {
        var apiUrl = (that.version >= 6) ? '/webapi/entry.cgi' : '/webapi/dsm/volume.cgi';
        var options = {
            url: that.url + apiUrl,
            qs: {
                api: (that.version >= 6) ? 'SYNO.Core.System' : 'SYNO.DSM.Volume',
                version: 1,
                method: (that.version >= 6) ? 'info' : 'list',
                type: 'storage', //only dsm >= 6
                _sid: that.auth.sid
            },
            method: 'GET',
            rejectUnauthorized: false,
            strictSSL: false
        };

        request(options, function(err, res, body) {
            if (!err && res.statusCode == 200) {
                var json = JSON.parse(body);

                //dsm version 5.x
                if (json.success && that.version < 6) {
                    var used = 0,
                        total = 0;
                    for (var i = json.data.volumes.length; i--;) {
                        used += json.data.volumes[i].used;
                        total += json.data.volumes[i].total;
                    }
                    callback(null, Math.round(used / total * 100));

                    //dsm version 6.x
                } else if (json.success && that.version >= 6) {
                    var used = 0,
                        total = 0;
                    for (var i = json.data.vol_info.length; i--;) {
                        used += json.data.vol_info[i].used_size;
                        total += json.data.vol_info[i].total_size;
                    }
                    callback(null, Math.round(used / total * 100));

                } else {
                    callback("An error occured while getting DiskUsage: " + json.error.code);
                }
            } else {
                callback("An error occured while getting DiskUsage: " + err);
            }
        });
    }).catch(function(err) {
        callback(err);
    });
};

module.exports = Synology;


src/SynologyAccessory.ts
import {
  AccessoryConfig,
  AccessoryPlugin,
  API,
  CharacteristicEventTypes,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
} from 'homebridge';
import PollingService from './PollingService';
import ping from 'ping';
import syno from 'syno';
import wol from 'wol';

interface RawQueryParams {
  sessionName: string;
  api: string;
  method: string;
  version: string;
  path: string;
}

interface DiskInfo {
  temp: number;
}

enum deviceStatus {
  Online = 'Online',
  Offline = 'Offline',
  WakingUp = 'WakingUp',
  ShuttingDown = 'ShuttingDown',
}

let hap: HAP;

export = (api: API) => {
  hap = api.hap;
  api.registerAccessory('synology', SynologyAccessory);
};

class SynologyAccessory implements AccessoryPlugin {

  private readonly log: Logging;
  private readonly api: API;
  private readonly name: string;
  private readonly host: string;
  private readonly mac: string;
  private readonly shutdownTime: number;
  private readonly startupTime: number;
  private readonly disabled: Array<string>;
  private readonly config: Record<string, unknown>;
  private dsm?: typeof syno;
  private readonly informationService: Service;
  private readonly switchService?: Service;
  private readonly temperatureService?: Service;
  private readonly diskTemperatureService?: Service;
  private state: deviceStatus;

  constructor(log: Logging, config: AccessoryConfig, api: API) {
    this.log = log;
    this.api = api;
    this.name = config.name;
    this.mac = config.mac;
    this.host = config.host;
    this.shutdownTime = config.shutdownTime || 60;
    this.startupTime = config.startupTime || 60;
    this.disabled = config.disabled || [];
    this.state = deviceStatus.Offline;
    this.config = {
      ignoreCertificateErrors: true,
      host: config.host,
      port: config.port || '5000',
      account: config.username,
      passwd: config.password,
      protocol: config.protocol,
      apiVersion: config.version || '6.2.2',
      otp: config.otp || undefined,
    };

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, 'Synology')
      .setCharacteristic(hap.Characteristic.Model, config.model || 'Diskstation')
      .setCharacteristic(hap.Characteristic.FirmwareRevision, config.version || '6.2.2')
      .setCharacteristic(hap.Characteristic.SerialNumber, config.serial || 'n/a');

    this.switchService = new hap.Service.Switch(this.name);
    this.switchService.getCharacteristic(hap.Characteristic.On)
      .on(CharacteristicEventTypes.SET, this.setState.bind(this));
    this.startStatePolling();

    if (!this.disabled.includes('temperature')) {
      this.temperatureService = new hap.Service.TemperatureSensor(`${this.name} Temperature`, 'systemp');
      this.startTemperaturePolling();
    }

    if (!this.disabled.includes('diskTemperature')) {
      this.diskTemperatureService = new hap.Service.TemperatureSensor(`${this.name} Average Disk Temperature`, 'disktemp');
      this.startDiskTemperaturePolling();
    }

    this.log.info(`Synology ${this.name} finished initializing!`);
  }

  startStatePolling(): void {
    const pollStateService = new PollingService(async () => {
      const res = await ping.promise.probe(this.host);
      pollStateService.emit('state', res.alive);
    }, 5000);

    pollStateService.on('state', (state: boolean) => {
      if ((this.state === deviceStatus.Online && state)
        || (this.state === deviceStatus.Offline && !state)
        || (this.state === deviceStatus.WakingUp && !state)
        || (this.state === deviceStatus.ShuttingDown && state)
      ) {
        return;
      }

      this.switchService!.updateCharacteristic(hap.Characteristic.On, state);
      this.state = state ? deviceStatus.Online : deviceStatus.Offline;
      this.log.info(`state changed to ${this.state}; updated characteristic`);
    });
    pollStateService.start();

    this.api.on('shutdown', (): void => pollStateService.stop());
  }

  startTemperaturePolling(): void {
    const pollTemperatureService = new PollingService(async () => {
      // don't poll temperature if device is not online
      if (this.state !== deviceStatus.Online) {
        this.temperatureService!.updateCharacteristic(hap.Characteristic.StatusActive, false);
        this.log.debug('Device is not online; Polling temperature disabled.');
        return;
      }

      try {
        const res = await this.query('dsm', 'getInfo');
        pollTemperatureService.emit('change', res.temperature);
      } catch (err) {
        this.temperatureService!.updateCharacteristic(hap.Characteristic.StatusActive, false);
        this.log.warn(`Can't get temperature, ${err}`);
      }
    }, 30000);

    pollTemperatureService.on('change', (temperature: number) => {
      this.log.debug(`updating temperature to ${temperature} °C`);
      this.temperatureService!.updateCharacteristic(hap.Characteristic.CurrentTemperature, temperature);
      this.temperatureService!.updateCharacteristic(hap.Characteristic.StatusActive, true);
    });

    pollTemperatureService.start();
    this.api.on('shutdown', (): void => pollTemperatureService.stop());
  }

  startDiskTemperaturePolling(): void {
    const pollDiskTemperatureService = new PollingService(async () => {
      // don't poll disk temperature if device is not online
      if (this.state !== deviceStatus.Online) {
        this.diskTemperatureService!.updateCharacteristic(hap.Characteristic.StatusActive, false);
        this.log.debug('Device is not online; Polling disk temperature disabled.');
        return;
      }

      try {
        const res = await this.rawQuery({
          sessionName: 'StorageInfo',
          api: 'SYNO.Storage.CGI.Storage',
          version: '1',
          path: 'entry.cgi',
          method: 'load_info',
        });
        const avgDiskTemp = (res.disks as DiskInfo[])
          .map((d: DiskInfo) => d.temp)
          .reduce((a: number, b: number) => a + b) / (res.disks as DiskInfo[]).length;
        pollDiskTemperatureService.emit('change', Math.round(avgDiskTemp * 10) / 10);
      } catch (err) {
        this.diskTemperatureService!.updateCharacteristic(hap.Characteristic.StatusActive, false);
        this.log.warn(`Can't get average disk temperature, ${err}`);
      }
    }, 30000);

    pollDiskTemperatureService.on('change', (temperature: number) => {
      this.log.debug(`updating average disk temperature to ${temperature} °C`);
      this.diskTemperatureService!.updateCharacteristic(hap.Characteristic.CurrentTemperature, temperature);
      this.diskTemperatureService!.updateCharacteristic(hap.Characteristic.StatusActive, true);
    });

    pollDiskTemperatureService.start();
    this.api.on('shutdown', (): void => pollDiskTemperatureService.stop());
  }

  async setState(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    // no change if device is waking up or shutting down
    if (this.state === deviceStatus.WakingUp || this.state === deviceStatus.ShuttingDown) {
      return callback();
    }

    // return if states matches
    if (this.state === deviceStatus.Online && value as boolean) {
      return callback();
    }

    if (value as boolean) {
      this.log.debug(`wake up ${this.name}.`);
      wol.wake(this.mac, (err) => {
        if (err) {
          this.log.warn(`Could not wake up ${this.name}: ${err}`);
          this.state = deviceStatus.Offline;
          return;
        }
        this.log.info(`${this.name} woke up!`);
        this.state = deviceStatus.WakingUp;
        setTimeout(() => {
          if (this.state !== deviceStatus.WakingUp) {
            return;
          }

          this.log.error(`Startup time of ${this.startupTime}s expired, reverting state to offline`);
          this.state = deviceStatus.Offline;
        }, this.startupTime * 1000);
      });
      callback();
    } else {
      try {
        this.log.info(`Shutting down ${this.name}`);
        await this.query('dsm', 'shutdownSystem');
        this.state = deviceStatus.ShuttingDown;
        this.dsm = undefined;
        setTimeout(() => {
          if (this.state !== deviceStatus.ShuttingDown) {
            return;
          }

          this.log.error(`Shutdown time of ${this.shutdownTime}s expired, reverting state to online`);
          this.state = deviceStatus.Online;
        }, this.shutdownTime * 1000);
      } catch(err) {
        this.log.error(`Can't shutdown ${this.name}: ${err}`);
        this.state = deviceStatus.Online;
      }
      callback();
    }
  }

  query(api: string, method: string, params:Record<string, unknown> = {}) {
    return new Promise<Record<string, unknown>>((resolve, reject) => {
      if (this.state !== deviceStatus.Online) {
        return reject(new Error(`${this.name} is offline, waking up or shutting down. No query possible`));
      }

      if (!this.dsm) {
        this.dsm = new syno(this.config);
      }

      this.dsm[api][method](params, (err: Record<string, unknown>, data: Record<string, unknown>) => {
        if (err) {
          return reject(err);
        }
        resolve(data);
      });
    });
  }

  rawQuery(params: RawQueryParams) {
    return new Promise<Record<string, unknown>>((resolve, reject) => {
      if (this.state !== deviceStatus.Online) {
        return reject(new Error(`${this.name} is offline, waking up or shutting down. No query possible`));
      }

      if (!this.dsm) {
        this.dsm = new syno(this.config);
      }

      this.dsm.dsm.requestAPI({
        params: null,
        done(err: Record<string, unknown>, data: Record<string, unknown>) {
          if (err) {
            return reject(err);
          }
          resolve(data);
        },
        apiInfos: params,
      });
    });
  }

  identify(): void {
    this.log(`Identifying Synology ${this.name}`);
  }

  getServices(): Service[] {
    const services: Service[] = [this.informationService];

    if (!this.disabled.includes('switch') && this.switchService) {
      services.push(this.switchService);
    }
    if (!this.disabled.includes('temperature') && this.temperatureService) {
      services.push(this.temperatureService);
    }
    if (!this.disabled.includes('diskTemperature') && this.diskTemperatureService) {
      services.push(this.diskTemperatureService);
    }

    return services;
  }
}
